room.cpp

```diff
@@ -826,6 +863,13 @@ void Room::processStateEvents(const RoomEvents& events)
                 }
                 break;
             }
+            case EventType::CallInvite:
+            case EventType::CallCandidates:
+            case EventType::CallAnswer:
+            case EventType::CallHangup: {
+              emit callEvent(this, event);
+              break;
+            }
```

Comment:
This is not the right place to catch `m.call` events - The Spec defines them as message events, so you have to put this code in `doAddNewMessages()` _and_ `doAddHistoricalMessages()` instead of from `processStateEvents()`. What's worse, in order for clients to correctly react to messages you have to do some extra checking:
- At least in `doAddHistoricalMessages()` you should make sure that clients are not notified about already finished calls. Note that historical message events come in reverse, hangup first, and moreover you might have a hangup event in one chunk and invite/candidates events in another chunk. `doAddNewMessages()` may benefit from the same logic but since call events come in natural order in this case, the standard logic on the client side will perfectly work.
- Since The Spec is very strict that `m.call` events should only be used in 1:1 rooms, it makes sense to only emit the signal if `memberCount() == 2`.

Reply:
One more thing here: by notifying about one of several call events with one common signal, you effectively mandate clients to put a `switch` statement in their code, to find out what exactly happened. Since that signal carries a `RoomEvent`, clients don't have single a chance to process the call event uniformly. So I'd really consider separating `callEvent()` into separate signals, maaybe only aggregating `CallInvite` and `CallCandidates` (and I'm not sure even that is proper).

Reply:
The reason why i choose `room` to place this event, is so client don't need to listen to each `room` to be able to catch incoming calls. This is even more useful if clients want add a "listen for call" background services for mobile devices.

But I do agree on 1:1 room check, I have done this on the client side for uMatriks but it does make sense to add this here, I'll add that tomorrow.

Reply:
I have no problem with `Room` being in the signal (although, unfortunately, you will anyway either have to put this signal to `Connection` - and listen for this signal from each connection object - or listen to `Room::callEvent()` from each room). I understand that calling is kind of a special case because it consumes much more attention of the end users (normally - all attention); so probably it really makes sense to make a separate signal on at least the level of `Connection`, or even setup a separate callback (although we mostly use direct connections that are only marginally heavier than callbacks).
My point in the original comment, though, was about passing a generic `RoomEvent` through a bottleneck of a single signal signature. When I get that `RoomEvent` in my slot on the client side, too many things have to be clarified before I can do anything sensible. Is that `RoomEvent` a call event at all (should I put a `Q_ASSERT(evt->isCallEvent())` in each client?)? Should I even care about a hangup event if I don't have active calls? Why should I care about (probably invalid) call answer event if I didn't place a call invite (note that I might have several devices and the call answer will arrive to all of them)? That's why I'd prefer to have 4 signals instead of a single one; and only `m.call.invite` is worth that "globalization" I mentioned in the first paragraph.